File: 39. Combination Sum/solution.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        sol = []
        if not candidates:
            return sol
        
        candidates.sort()
        sol = self.backtrack(candidates, target, 0)
        for s in sol:
            s.reverse()
            
        return sol
        
    def backtrack(self, candidates, target, n):
        sol = []
        for i in range(n, len(candidates)):
            sub = []
            if target - candidates[i] < 0:
                break
            elif target - candidates[i] == 0:
                sub.append([candidates[i]])
            else:
                sub = self.backtrack(candidates, target-candidates[i], i)
                for s in sub:
                    s.append(candidates[i])
            if sub:
                sol.extend(sub)
                
        return sol
            

File: 39. Combination Sum/solution.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        Arrays.sort(candidates);
        List<List<Integer>> sol = new ArrayList<List<Integer>>();
        
        List<Integer> subsol = new ArrayList<Integer>();
        backtrack(candidates, target, 0, sol, subsol);
        return sol;
    }
    
    public void backtrack(int[] candidates, int target, int pos, List<List<Integer>> sol, List<Integer> subsol) {
        if (target == 0) {
            sol.add(new ArrayList<Integer>(subsol));
            return;
        }
        
        for (int i = pos; i < candidates.length; i++) {
            if (target-candidates[i] < 0) {
                return;
            }
            subsol.add(candidates[i]);
            backtrack(candidates, target-candidates[i], i, sol, subsol);
            subsol.remove(subsol.size()-1);
        }
    }
}
